File: leetcode/C++/course-schedule-iii.cpp
class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(),courses.end(),[](vector<int> &a, vector<int> &b){return a[1]<b[1];});
        int cur=0;
        priority_queue<int> pq;
        for(int i=0; i<courses.size(); ++i){
            cur+=courses[i][0];
            pq.push(courses[i][0]);
            if(cur>courses[i][1]){
                cur-=pq.top();
                pq.pop();
            }
        }
        return pq.size();
    }
};

File: leetcode/Python/course-schedule-iii.py
from queue import PriorityQueue
class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        cur, pq = 0, PriorityQueue()
        for course in sorted(courses, key= lambda x: x[1]):
            cur+=course[0]
            pq.put(-course[0])
            if cur > course[1]:
                cur+=pq.get()
        return pq.qsize()